timemachine/cpp/src/kernels/k_sampling.cu

```diff
@@ -38,6 +38,67 @@ template void __global__ k_setup_gumbel_max_trick<double>(
     const double *__restrict__ gumbel_noise,
     double *__restrict__ prepared_gumbel);
 
+template <typename RealType>
+void __global__ k_setup_gumbel_max_trick_with_offset(
+    const int num_segments,
+    const int total_values,
+    const int max_offset,
+    const int *__restrict__ noise_offset,     // [1]
+    const int *__restrict__ segment_offsets,  // [num_segments]
+    const RealType *__restrict__ log_weights, // [total_values]
+    const RealType *__restrict__ gumbel_noise,
+    RealType *__restrict__ prepared_gumbel) {
+    const int segment_idx = blockIdx.y;
+
+    if (segment_idx >= num_segments) {
+        return;
+    }
+
+    const int rand_offset = noise_offset[0];
+    if (rand_offset + segment_idx >= max_offset) {
+        return;
+    }
+
+    const int segment_start = segment_offsets[segment_idx];
+    const int N = segment_offsets[segment_idx + 1] - segment_start;
+    // In the case of the offset the values per segment need to match up to compute gumbel offset correctly.
+    assert(total_values % N == 0);
+
+    const int gumbel_offset = rand_offset * N + segment_start;
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    while (idx < N) {
+        const RealType weight = log_weights[idx];
+        assert(!isnan(weight));
+
+        // If the idx in the batch segments is beyond the max offset return a negative infinity to avoid sampling the weight
```

Comment:
is this comment still relevant? Not clear where/how the negative infinity is returned

Reply:
Nope, will remove

Reply:
Removed in https://github.com/proteneer/timemachine/pull/1274/commits/6bf2b83d9185f956eb88558cce6671b56bdb5918